File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.java
class Solution {
    class polak{
        int first;
        int second;
        int third;
        polak(int f,int s,int t){
            first=f;
            second=s;
            third=t;
        }
    }
    public int shortestPathBinaryMatrix(int[][] grid) {
        int n=grid.length;
        int m=grid[0].length;
        int[][] dist=new int[n][m];
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                dist[i][j]=(int)1e9;
            }
        }
        if(grid[0][0]==1 || grid[n-1][m-1]==1){
            return -1;
        }
        if(n==1 && m==1){
            return 1;
        }
        
        dist[0][0]=1;
        Queue<polak> q=new LinkedList<>();
        q.add(new polak(1,0,0));
        while(!q.isEmpty()){
            int dis=q.peek().first;
            int row=q.peek().second;
            int col=q.peek().third;
            q.poll();
            int drow[]={-1,-1,-1,0,0,+1,1,1};
            int dcol[]={-1,0,1,-1,+1,-1,0,1};
            for(int i=0;i<8;i++){
                int newrow=row+drow[i];
                int newcol=col+dcol[i];
                if(newrow>=0 && newcol>=0 && newrow<n && newcol<m && grid[newrow][newcol]==0 && dis+1<dist[newrow][newcol]){
                    dist[newrow][newcol]=dis+1;
                    if(newrow==n-1 && newcol==m-1){
                        return dis+1;
                    }
                    q.add(new polak(dis+1,newrow,newcol));
                }
            }
        }
        return -1;
    }
}

File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.cpp
class Solution {
public:
    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
     int n=grid.size();
     int m=grid[0].size();
     if(grid[0][0] ==1 || grid[n-1][m-1]==1){
        return -1;
     }
     priority_queue<tuple<int,int,int>, vector<tuple<int,int,int>>, greater<tuple<int,int,int>>> q;
     int dx[]={1,0,-1,0,1,-1,1,-1};
     int dy[]={0,1,0,-1,1,-1,-1,1};
     q.push({0,0,1});
    
     vector<vector<int>> dist(n,vector<int>(m,INT_MAX));
     dist[0][0]=1;
     while(!q.empty()){
        auto[x,y,cnt]=q.top();
        
        q.pop();
        if(x==n-1 && y==m-1){
            return cnt;
        }
        for(int i=0;i<8;i++){
            int newrow=x+dx[i];
            int newcol=y+dy[i];
            if(newrow>=0 && newcol >=0 && newrow<n && newcol <m && grid[newrow][newcol]==0){
                if(cnt+1<dist[newrow][newcol]){
                    dist[newrow][newcol]=cnt+1;
                    q.push({newrow,newcol,cnt+1});
                }
                    
                    
            }
        }
     }
     return -1;
    }
};